### DISCUSSION TITLE
Postgres error - could not connect to server: No such file or directory

### DISCUSSION THREAD

**Discussion by thornomad** [community/community#25137] [Actions]
<p>Hello - hoping someone can point me in the right direction.  Trying to get an action working for Rails with Postgres support.  Got the authorization bit sorted out (finally!) but now am at this error:</p>
<pre><code>could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
Couldn't create 'api_test' database. Please check your configuration.
rake aborted!
ActiveRecord::NoDatabaseError: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
</code></pre>
<p>This looks vaguely familiar to me.  I had an issue like this with the Docker image on my local machine and I ended up have to specify a <code>volume</code> in my docker_compose.yml file.  I think normally it has to do with an issue with the database file on postgres not being setup right.  Feel like I am following the instructions … so wonder what I am missing?</p>
<p>Here is my ruby.yml configuration setup:</p>
<pre><code>name: Ruby

on: [push]

jobs:
  build:

    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:12.2
        env:
          POSTGRES_PASSWORD: postgres
          POSTGRES_USER: postgres
        ports:
          - 5432:5432
        options: &gt;-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5

    steps:
    - uses: actions/checkout@v2

    - uses: actions/cache@v1
      with:
        path: vendor/bundle
        key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile.lock') }}
        restore-keys: |
          ${{ runner.os }}-gems-

    - name: Set up Ruby 2.7.1
      uses: ruby/setup-ruby@v1.30.1
      with:
        ruby-version: 2.7.1

    - name: Install PostgreSQL Client
      run: |
        sudo apt-get -yqq install libpq-dev

    - name: Run bundle install
      env:
        RAILS_ENV: test
      run: |
        bundle config path vendor/bundle
        bundle install --jobs 4 --retry 3

    - name: Build and test with rspec
      env:
        RAILS_ENV: test
        POSTGRES_HOST: localhost
        POSTGRES_USER: postgres
        POSTGRES_PASSWORD: postgres
        POSTGRES_PORT: ${{ job.services.postgres.ports[5432] }}
        RUBYOPT: "-W:no-deprecated -W:no-experimental"
      run: |
        bundle exec rake db:setup
        bundle exec rspec
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/thornomad">@thornomad</a> ,</p>
<p>In the last step, you can access the service container using localhost and the mapped port directly.</p>
<pre><code>POSTGRES_PORT: 5432
</code></pre>
<p>Please refer to the doc <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idservices" rel="nofollow noopener">here</a> for more details.</p>
<p>If it doesn’t work, please kindly share your repo/sample repo for further investigation.</p>
<p>Thanks.</p>

**Comment by thornomad** [comment]
<p>Hello <a class="mention" href="/weide-zhou">@weide-zhou</a> - thanks for taking time to help me.  I have another repo I had used for a different test (basic rails app that uses postgres with no real configuration).  I added the github action to that app in its own branch to demonstrate the failure I am having.  You can see the action setting at:</p>
<p><a href="https://github.com/thornomad/issue-38279-rails-6/blob/add-github-action/.github/workflows/ruby.yml" rel="nofollow noopener">https://github.com/thornomad/issue-38279-rails-6/blob/add-github-action/.github/workflows/ruby.yml</a></p>
<p>I am taking a guess at the config/database.yml setup.  I added the password, username, and port evn variables there.</p>
<p><a href="https://github.com/thornomad/issue-38279-rails-6/blob/add-github-action/config/database.yml#L58-L63" rel="nofollow noopener">https://github.com/thornomad/issue-38279-rails-6/blob/add-github-action/config/database.yml#L58-L63</a></p>
<p>Finally, here is the output from the action.  Am having the same issue as with the other application.</p>
<p><a href="https://github.com/thornomad/issue-38279-rails-6/runs/564539190?check_suite_focus=true" rel="nofollow noopener">https://github.com/thornomad/issue-38279-rails-6/runs/564539190?check_suite_focus=true</a></p>
<p>Thanks again.  I feel like all my searching online and trying different examples floating around and I am getting stuck.  Must be something fundamental that I am missing!</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/thornomad">@thornomad</a> ,</p>
<p>Thanks for your reply! I forked your repo and did some changes, now the error is gone:</p>
<ol>
<li>
<p>In config/database.yml, change to use below code, use ‘adapter’ &amp; ‘host’ parameter:</p>
<p>test:<br>
&lt;&lt;: *default<br>
database: test_app_6_test<br>
adapter: postgresql<br>
host: localhost<br>
encoding: unicode<br>
username: &lt;%= ENV[“POSTGRES_USER”] %&gt;<br>
password: &lt;%= ENV[“POSTGRES_PASSWORD”] %&gt;</p>
</li>
<li>
<p>The workflow file:</p>
<p>name: Ruby</p>
<p>on: [push]</p>
<p>jobs:<br>
build:<br>
runs-on: ubuntu-latest<br>
services:<br>
postgres:<br>
image: postgres:11<br>
ports:<br>
- 5432:5432<br>
env:<br>
POSTGRES_USER: postgres<br>
POSTGRES_PASSWORD: postgres<br>
options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5</p>
<pre><code> steps:
   - uses: actions/checkout@v1

   - uses: actions/setup-ruby@v1
     with:
       ruby-version: 2.6.x

   - name: Install dependent libraries
     run: sudo apt-get install libpq-dev

   - name: Bundle install
     run: |
       gem install bundler
       bundle install --jobs 4 --retry 3
   - name: setup databases
     env:
       POSTGRES_PASSWORD: postgres
       POSTGRES_USER: postgres
       RAILS_ENV: test
     run: bundle exec rake db:setup
</code></pre>
</li>
<li>
<p>In the Gemfile, i changed ruby version from ‘2.6.3’ to ‘2.6.5’, since verion 2.6.3 cannot be found by action/setup-ruby.</p>
</li>
</ol>
<p>My workflow run here: <a href="https://github.com/weide-zhou/issue-38279-rails-6/runs/566774300?check_suite_focus=true" rel="nofollow noopener">https://github.com/weide-zhou/issue-38279-rails-6/runs/566774300?check_suite_focus=true</a></p>
<p>You can refer to the <a href="https://www.pibit.nl/github/actions/rails/postgres/rspec/tutorial/example/2019/09/23/github-actions-with-rails-postgres-and-rspec/" rel="nofollow noopener">link</a> for more details. Hope it helps!</p>

**Comment by thornomad** [comment]
<p>Hi - thanks for working with me on this.  It looks like the issue in the end was not defining <code>host: localhost</code> in the <code>database.yml</code> file.</p>
<p>For anyone else who stubmles across this; here is what ended up working for me on a new Rails 6 setup using postgres.  This let’s both the local machine testing environment and github actions do its thing! Thanks again!</p>
<pre><code>env:
  POSTGRES_PASSWORD: postgres
  POSTGRES_USER: postgres
  POSTGRES_HOST: localhost
  RAILS_ENV: test

name: Ruby

on: [push]

jobs:
  build:
    runs-on: ubuntu-latest
    services:
      postgres:
        image: postgres:latest
        ports:
          - 5432:5432
        env:
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: postgres
        options: &gt;-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5

    steps:
      - uses: actions/checkout@v2

      - name: Set up Ruby version (using project .ruby-version)
        uses: eregon/use-ruby-action@master

      - name: Install dependent libraries
        run: sudo apt-get install libpq-dev

      - name: Bundle install
        run: |
          gem install bundler
          bundle install --jobs 4 --retry 3

      - name: Setup the postgres database
        run: bundle exec rake db:setup

      - name: Run RSpec tests
        run: bundle exec rspec
</code></pre>
<p> And then the relevant section from <code>database.yml</code>:</p>
<pre><code>test:
  &lt;&lt;: *default
  database: my_app_test
  host: &lt;%= ENV["POSTGRES_HOST"] %&gt;
  username: &lt;%= ENV["POSTGRES_USER"] %&gt;
  password: &lt;%= ENV["POSTGRES_PASSWORD"] %&gt;
</code></pre>
<p>Thanks again!</p>
